### ISSUE TITLE
Supporting tweets?

### ISSUE THREAD

**Issue by michalrus** [Sorunome/mx-puppet-twitter#9]
Hi!

The old bridge used to do that @ https://half-shot.github.io/matrix-appservice-twitter/ 

Would that not be a nice idea?

**Comment by Sorunome** [comment]
How do you imagine this to work? A single room where all tweets get dumped in?

**Comment by IngwiePhoenix** [comment]
> How do you imagine this to work? A single room where all tweets get dumped in?

Essentially, yes.

I was wondering about a similiar feature myself when I saw that the previous bridge had been set to archived, which put me well off from using it, since this made it clear that that bridge would no longer be maintained. And that is how I came here :)

For a long time, before the Stream API was moved to the "pro" developer subscription tier, I used a tool called Rainbowstream: https://github.com/orakaro/rainbowstream . I couldn't find the issue ticket, but we were actually discussing possible alternatives in there already - and one was periodically requesting the newest tweets. Now, since Matrix is HTTP based, this is ironiically fitting as HTTP is not a constant stream anyway - so the user wouldn't even notice a delay anyway. And this is basically where my idea starts.

What I would imagine being doable is that the bridge can create a room for all tweets, mentions and notifications (or a similiar configuration - this is really just an idea after all but i disgress). Within those, the user could reply to the messages, presented as tweets, to have an answer tweeted. This is definitively do-able, as a `m.room.message` event can have fitting reply parameters added ("Rich replies", linked below). That way, the bridge would be able to tell what tweet corresponds to the message being replied to, and thus could instantiate the tweet-reply.

In order to follow a tweet, the user might also be able to copy a tweet's ID. That ID could also be part of the message, which is how we did things in Rainbowstream, where an ID was shown which we could adress. Then again, that was a CLI application, so something more useful might be better for Matrix. I don't know if it has links that cause a room to be generated or something. 

Another thing that could be done would be to allow someone to spawn a private channel which was simply dedicated to showing tweets by a specific user,

I am still cross-reading the Matrix API to find possibilities (and boy are they endless...this documentation is huge!) for what could possibly be done. But i think just having a channel that shows all the tweets from followed users would be a nice start. The second most important one would be mentions and general notifications. The tweet-thread or specific-user-tweets channel would be of a lesser importance, but probably not impossible either. The most difficult things would probably be replying to tweets and opening a whole tweet thread.

What do you think about this? :) I personally really love Matrix and having my Twitter integrated to it would be pretty awesome! And being able to see right away when someone specifically posts something new would also be quite helpful. I have been using TweetDeck for a long while now, which is where the idea of multiple channels originates from. Also, the fact that Matrix is _kind of_ a glorified AJAX-based chat can actually become quite helpful here. Yes, it's not quite like ajax-based chats once were like, but it is still similiar as to using HTTP exclusively.

Anyway, I hope you have a great day and maybe you can take away some ideas from here! ^.^

* Rich replies: https://matrix.org/docs/spec/client_server/latest#rich-replies

**Comment by michalrus** [comment]
Wow, basically what @IngwiePhoenix said! :heart: 

I would also love to be able to define:

* rooms following a certain #hashtag, e.g. https://twitter.com/hashtag/Haskell

**Comment by Sorunome** [comment]
Yes, these ideas sound good! Problem is, that soru is only one person with limited time and a high workload, so no idea when she'll actually get to it ^^

She vaguely recalls the webhook API allowing to receive tweet notifications, so polling wouldn't be needed for that.

So yeah, it's on the stack of things to do, but if someone else wants to step up and implement this (with help?) it would also be awesome!

Am 10. März 2020 16:12:02 MEZ schrieb Ingwie Phoenix <notifications@github.com>:

**Comment by TheLlamaLover** [comment]
With the upcoming changes to the Twitter API as mentioned in https://developer.twitter.com/en/docs/twitter-api/early-access, could tweets be supported?
